src/shared_memory_manager.h

```diff
@@ -130,14 +106,80 @@ class SharedMemoryManager {
   /// \return a TRITONSERVER_Error indicating success or failure.
   TRITONSERVER_Error* UnregisterAll(TRITONSERVER_MemoryType memory_type);
 
+#ifdef TRITON_ENABLE_GPU
+  /// Add a shared memory block representing shared memory in CUDA
+  /// (GPU) memory to the manager. Return TRITONSERVER_ERROR_ALREADY_EXISTS
+  /// if a shared memory block of the same name already exists in the manager.
+  /// \param name The name of the memory block.
+  /// \param cuda_shm_handle The unique memory handle to the cuda shared
+  /// memory block.
+  /// \param byte_size The size, in bytes of the block.
+  /// \param device id The GPU number the shared memory region is in.
+  /// \return a TRITONSERVER_Error indicating success or failure.
+  TRITONSERVER_Error* RegisterCUDASharedMemory(
+      const std::string& name, const cudaIpcMemHandle_t* cuda_shm_handle,
+      const size_t byte_size, const int device_id);
+
+  /// Get the CUDA memory handle associated with the block name.
+  /// Return TRITONSERVER_ERROR_NOT_FOUND if named block doesn't exist.
+  /// \param name The name of the shared memory block to get.
+  /// \param cuda_mem_handle Returns the cuda memory handle with the memory
+  /// block.
+  /// \return a TRITONSERVER_Error indicating success or failure.
+  TRITONSERVER_Error* GetCUDAHandle(
+      const std::string& name, cudaIpcMemHandle_t** cuda_mem_handle);
+#endif  // TRITON_ENABLE_GPU
+
  private:
   /// A helper function to remove the named shared memory blocks of
   /// specified type
   TRITONSERVER_Error* UnregisterHelper(
       const std::string& name, TRITONSERVER_MemoryType memory_type);
 
-  /// A struct that records the shared memory regions registered by the shared
-  /// memory manager.
+/// A struct that records the shared memory regions registered by the shared
+/// memory manager.
+#ifdef _WIN32
+  struct SharedMemoryInfo {
```

Comment:
My intuition says that there is a better way to do this, but I don't have a solution atm. The only way these structs differ is the file `fd`/`HANDLE` which we hold onto. They are both "handles" in that they are both integers which we ask the operating system about and act on.  Should we alias/typedef the struct, make a union, something else? As a starting point maybe something like:

```
#ifdef _WIN32
  typedef HANDLE file_handle;
#else
  typedef int file_handle;
#endif
```

This would be defined at the top of the file so that we can replace the type of the `shm_handle_` as `file_handle` and only define the struct once. I'd ask around on this to make sure I'm not crazy.

Reply:
Yes, in fact, that is exactly what I currently have implemented on the client side.

Tanmay's feedback was to make a substruct containing platform-dependent members. Currently it would only manage int vs HANDLE, but could be extended in the future. It would look something like this:
```c++
struct ShmFile {
#ifdef _WIN32
  HANDLE shm_file_;
  ShmFile(HANDLE shm_file): shm_file_(shm_file) {};
#else
  int platform_handle_;
  ShmFile(int shm_file): shm_file_(shm_file) {};
#endif // _WIN32
};
```
where `SharedMemoryInfo`'s `shm_handle` member would be a pointer to a ShmFile object.